implementation_details.md

```diff
@@ -0,0 +1,160 @@
+
+Again, people often imagine having the server estimate what interpolated state the client was looking at based on their RTT, but we can resolve this without any guesswork.
+
+Clients can just tell the server what they were looking at by bundling the interpolated tick numbers and the blend value inside the input payloads. With this information, the server can reconstruct *exactly* what each client saw.
+
+```
+<packet header>
+tick number (predicted)
+tick number (interpolated from)
+tick number (interpolated to)
+interpolation blend value
+<rest of payload>
+```
+
+So there are two ways to go about the actual compensation:
+- Compensate upfront by bringing new projectiles into the present (similar to a rollback).
+- Compensate over time ("amortized"), constantly testing projectiles against the history buffer.
+
+There's a lot to learn from *Overwatch* here.
+
+*Overwatch* shows that [time is just another collision dimension](https://youtu.be/W3aieHjyNvw?t=2226). Basically, you can broadphase test against the entire collider history at once (with the amortized method).
+
+*Overwatch* [allows defensive abilities to mitigate compensated projectiles](https://youtu.be/W3aieHjyNvw?t=2492). AFAIK this is simple to do. If a player activates any defensive bonus, just apply it to all their buffered hitboxes.
+
+For clients with too-high ping, their interpolation will lag far behind their prediction. If you only compensate up to a limit (e.g. 200ms), [those clients will have to extrapolate the difference](https://youtu.be/W3aieHjyNvw?t=2347). Doing nothing is also valid, but lagging clients would abruptly have to start leading their targets.
+
+## Unconditional Rollbacks
+Every article on "rollback netcode" and "client-side prediction and server reconciliation" encourages having clients compare their predicted state to the authoritative state and reconciling *if* they mispredicted. But how do you actually detect a mispredict?
+
+I thought of two methods while I was writing this:
+
+1. Unordered scan looking for first difference.
+2. Ordered scan to compute checksum and compare.
+
+The first option has an unpredictable speed. The second option requires a fixed walk of the game state (checksums *are* probably worth having even if only for debugging non-determinism). There may be options I didn't consider, but the point I'm trying to make is that detecting changes among large numbers of entities isn't cheap. 
+
+Let's consider a simpler default:  
+
+3. Always rollback and re-simulate.
+
+Now, if you're thinking that's wasteful, the "if mispredicted" gives you a false sense of security. If I make a game and claim it can rollback 250ms, that basically should mean *any* 250ms, with no stuttering. If clients *always* rollback and re-sim, it'll be easier to profile and optimize for that. As a bonus, clients never need to store old predicted states.
```

Comment:
Good point.

I just wanted to clarify this bit:

> As a bonus, clients never need to store old predicted states.

It means that we still store the history of authoritative updates, but clients can simply avoid adding predicted states on top of those, correct? And we still need to store a buffer of local players' commands, to be able to re-sim the predicted states.

Reply:
Yes, that's exactly it.